pkg/kubelet/util/pod_startup_latency_tracker.go

```diff
@@ -0,0 +1,180 @@
+	if err != nil {
+		klog.ErrorS(err, "Unable to parse", "pod.ResourceVersion", pod.ResourceVersion)
+		return
+	}
+
+	if podResourceVersion >= state.firstResourceVersionWhenPodWasSeenAsRunning {
+		podStartingDuration := when.Sub(pod.CreationTimestamp.Time)
+		imagePullingDuration := state.lastFinishedPulling.Sub(state.firstStartedPulling)
+		podStartSLOduration := (podStartingDuration - imagePullingDuration).Seconds()
+
+		klog.V(2).InfoS("Observe PodStartSLIDuration metric", "pod", klog.KObj(pod), "podStartSLOduration", podStartSLOduration, "watchObservedTime", when, "pod.CreationTimestamp", pod.CreationTimestamp.Time, "firstStartedPulling", state.firstStartedPulling, "lastFinishedPulling", state.lastFinishedPulling)
+
+		metrics.PodStartSLIDuration.WithLabelValues().Observe(podStartSLOduration)
+		state.recorded = true
+	}
+}
+
+func (p *PodStartupLatencyTracker) RecordImageStartedPulling(podUID types.UID) {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	state := p.pods[podUID]
+	if state == nil {
+		klog.ErrorS(nil, "ObservedPodOnWatch should happened before image pulling", "pod", podUID)
+		return
+	}
+
+	if state.firstStartedPulling.IsZero() {
+		state.firstStartedPulling = p.clock.Now()
+	}
+}
+
+func (p *PodStartupLatencyTracker) RecordImageFinishedPulling(podUID types.UID) {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	state := p.pods[podUID]
+	if state == nil {
+		klog.ErrorS(nil, "ObservedPodOnWatch should happened before image pulling", "pod", podUID)
+		return
+	}
+
+	state.lastFinishedPulling = p.clock.Now() // Now is always grater than values from the past.
+}
+
+func (p *PodStartupLatencyTracker) RecordStatusUpdated(pod *v1.Pod) {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	state := p.pods[pod.UID]
+	if state == nil {
+		klog.ErrorS(nil, "ObservedPodOnWatch should happened before status changes", "pod", klog.KObj(pod))
+		return
+	}
+
+	if state.recorded {
+		// skip, pod latency already recorded
+		return
+	}
+
+	if state.firstResourceVersionWhenPodWasSeenAsRunning != 0 {
+		// skip, pod already started
+		return
+	}
+
+	if hasPodStartedSLO(pod) {
+		version, err := strconv.ParseUint(pod.ResourceVersion, 10, 64)
+		if err != nil {
+			klog.ErrorS(err, "Unable to parse", "pod.ResourceVersion", pod.ResourceVersion)
+			return
+		}
+
+		klog.V(2).InfoS("Mark when the pod was running for the first time", "pod", klog.KObj(pod), "rv", version)
+		state.firstResourceVersionWhenPodWasSeenAsRunning = version
+	}
+}
+
+// hasPodStartedSLO should reflect "Pod startup latency SLI" definition
+// ref: https://github.com/kubernetes/community/blob/master/sig-scalability/slos/pod_startup_latency.md
+func hasPodStartedSLO(pod *v1.Pod) bool {
+	if pod.Status.Phase == v1.PodSucceeded {
+		return true
+	}
+	if pod.Status.Phase == v1.PodRunning && podReady(pod) {
+		return true
+	}
+	return false
+}
+
+func podReady(pod *v1.Pod) bool {
+	for _, cond := range pod.Status.Conditions {
+		if cond.Type == v1.PodReady && cond.Status == v1.ConditionTrue {
```

Comment:
Quick question here -- the pod will only be reported Ready if the containers are reported as ready. See logic here - https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/status/generate.go#L105

So for example, if a given container had a probe set that never went Ready the container would not be ready and as such the pod would not be ready. This is of course application specific, since the workload needs to startup and become ready.

This seems a bit conflict with the goal mentioned on https://github.com/kubernetes/community/blob/master/sig-scalability/slos/pod_startup_latency.md:

> we don't want to rely on "readiness checks", because they heavily depend on the application. If the application takes couple minutes to initialize before it starts responding to readiness checks, that shouldn't count towards Kubernetes performance.

Does the current SLO-monitor also look at pod ready condition?

Reply:
Yes, it looks at the pod ready condition. I've copied the logic from slo-monitor (+some code refactoring): https://github.com/kubernetes/perf-tests/blob/master/slo-monitor/src/monitors/util.go#L29-L60

But this is a good point - if we want to follow the definition, rather than replicate `slo-monitor`, we probably should change it.
```
pod.Status.ContainerStatuses ".every"
 - option 1) .Started
 - option 2) .State.Running.StartedAt.IsNotZero()
```

Picking between (1) and (2) depends on, how do we want to tread restarted containers? Do we want to capture the moment when all the containers are running vs every container has started.

Reply:
I think we should wait until each container has been started at least once.

Reply:
hasPodStartedSLO func updated